p2p/raintree/network.go

```diff
@@ -26,8 +26,11 @@ import (
 var (
 	_ typesP2P.Network           = &rainTreeNetwork{}
 	_ modules.IntegratableModule = &rainTreeNetwork{}
+	_ rainTreeFactory            = &rainTreeNetwork{}
 )
 
+type rainTreeFactory = modules.FactoryWithConfig[typesP2P.Network, RainTreeConfig]
```

Comment:
so we're going to have to do this everywhere, right?

If so, can you add a techdebt comment where with define `FactoryWithConfig` et al to not forget?

Reply:
Could you elaborate on "this" and "everywhere"? If you're asking about defining additional types and/or enforcement using the generic factory interfaces, that would only be necessary for "sub-modules" as the `ModuleFactoryWithOptions` interface is currently embedded in the `Module` interface.

Reply:
> "this"

```go
type xxxFactory = modules.FactoryWithConfig[XXXModule,XXXConfig]
```

>  "everywhere"

Everywhere where we create a module

> If you're asking about defining additional types and/or enforcement using the generic factory interfaces, that would only be necessary for "sub-modules" as the ModuleFactoryWithOptions interface is currently [embedded in the Module interface](https://github.com/pokt-network/pocket/blob/main/shared/modules/module.go?rgh-link-date=2023-04-24T06%3A16%3A26Z#L11).

Noted, this answers my question above.

With that said the, two follow-up:

1. Can you remind me why `Module` embedding `FactoryWitOptions` `FactoryWithConfigAndOptions`
2. So are we going to find "all" the places where have to do "this" where we have existing submodules (e.g. StateSync)?

Reply:
Thanks for the clarification! :sweat_smile: 

> 1. Can you remind me why Module embedding FactoryWitOptions FactoryWithConfigAndOptions

`Module` is embedding `ModuleFactoryWithOptions` (which is an alias for `FactoryWithOptions[Module, ModuleOptions]`) because that was consistent with the existing convention and existing module implementations already enforce the `Module` interface. By embedding it there, we didn't need to do any additional enforcement for modules.

I don't think there's a legitimate case where a module would use anything other than `ModuleFactoryWithOptions` because modules are conventionally get configs from the runtime module via the bus. We could of course use this same mechanism for submodules but I would expect that in most cases, forcing sub-modules to use the bus for configs would lead to unnecessary additional complexity.

> 2. So are we going to find "all" the places where have to do "this" where we have existing submodules (e.g. StateSync)?

Yes. Comments added. :raised_hands:

Reply:
Thanks for the clarification! :sweat_smile: 

> 1. Can you remind me why Module embedding FactoryWitOptions FactoryWithConfigAndOptions

`Module` is embedding `ModuleFactoryWithOptions` (which is an alias for `FactoryWithOptions[Module, ModuleOptions]`) because that was consistent with the existing convention and existing module implementations already enforce the `Module` interface. By embedding it there, we didn't need to do any additional enforcement for modules.

I don't think there's a legitimate case where a module would use anything other than `ModuleFactoryWithOptions` because modules are conventionally get configs from the runtime module via the bus. We could of course use this same mechanism for submodules but I would expect that in most cases, forcing sub-modules to use the bus for configs would lead to unnecessary additional complexity.

> 2. So are we going to find "all" the places where have to do "this" where we have existing submodules (e.g. StateSync)?

Yes. Comments added. :raised_hands: